### ISSUE TITLE
Kafka message bus does not support module.<name>.count=0 

### ISSUE THREAD

**Issue by mehiel** [spring-projects/spring-xd#1593]
@mbogoevici following up our short twitter conversation it seems that kafka message bus doesn't support module.<name>.count=0 for direct binding and that should be better documented along with different solutions for direct binding in that case.

**Comment by mehiel** [comment]
TBH I don't really like the count=0 way of doing direct binding but I got there as a last resort, because my YARN deployment has problems with composed modules (throws a `FileNotFoundException`or something similar, I will create another issue on that) and for various reasons I want two modules to be directly bound.

And I'd really like to know if there're other ways of doing direct binding other than count=0 and composition.

**Comment by mbogoevici** [comment]
@mehiel - thanks for reaching out, here's a bit of clarification

1) The Kafka message bus doesn't support module.x.count=0, which implies the ability to increase/decrease the module count automatically with the arrival of a new container, and automatically reassign partitions to modules. This is mainly a feature of the high lever consumer, which we've abandoned due to its issues and limitations (see https://cwiki.apache.org/confluence/display/KAFKA/Consumer+Client+Re-Design and also http://www.michael-noll.com/blog/2014/10/01/kafka-spark-streaming-integration-example-tutorial/#kafka-integration for details). It is not a common approach in other environments that use Kafka as transport either (see Apache Samza, with the notable difference that a Spring XD module may listen to more than 1 Kafka partition - still deterministic). Also, ensuring that partitions are deterministically assigned to modules and the allocation doesn't change has benefits with regards to stateful processing and data affinity. We've found module.x.count=0 to be a better suited approach for MOM's like Rabbit, which allows a competing consumer model.

tl;dr We're not supporting module.x.count=0 because we prefer deterministic partition allocations with Kafka and we want to avoid the issues arising from rebalancing, as well as the use of the high level consumer. Should there be a need for it, we will reconsider our approach. 

2) I think that one quick way to fix the issues that you are encountering is to fix module composition in YARN. Please file a JIRA (https://jira.spring.io/browse/XD) issue for that and we will address it as soon as possible. 

3) Using direct binding solely as a side effect of module.*.count=0 is not a path that we intend to pursue on the long term. In fact, we've identified this as an issue/missing feature, and we intend to fix it via https://jira.spring.io/browse/XD-2946. Please vote on the issue to ensure it gets a higher priority on the current roadmap. 

4) We will properly document count=0 as an intentional limitation of the Kafka bus. 

I hope this helps, and thanks for using Spring XD!

**Comment by mbogoevici** [comment]
Also, while we will get to this as soon as possible, we are always happy to receive contributons to speed things up :)

**Comment by mbogoevici** [comment]
@mehiel just a quick update on this: ever since we've had this discussion, we've released Spring XD 1.2: https://spring.io/blog/2015/06/17/spring-xd-1-2-ga-spring-xd-1-1-3-and-flo-for-spring-xd-beta-released

While we couldn't actually get around supporting https://jira.spring.io/browse/XD-2946, module composition in YARN should be now fixed: https://jira.spring.io/browse/XD-3171 so perhaps it's worth giving it a try again.

We've also documented that count=0 is not a supported feature of the Kafka bus.

I'll close the issue for now, but feel free to comment and re-open if you think that there's anything outstanding.

**Comment by mehiel** [comment]
Many thanks @mbogoevici 

Yes, having composed modules on YARN and documentating about count=0 and Kafka is more than enough to close this one. I'll update to 1.2 asap.

I saw you guys released an Ambari plugin too. I'm a huge fan of Ambari but I think I still prefer XD on YARN. Probably having a plugin that works with XD on YARN would be greak. Giving insights, metrics and other functionality from bin/xd-yarn probably. 

But this is a discussion for another topic anyways.
